File: src/color.ts
import { Vec3 } from "./vec3";

const n = (m: number) => Math.floor(m * 255);

const linearToGamma = (linearComponent: number) => {
  return linearComponent > 0 ? Math.sqrt(linearComponent) : 0;
};

function fillRGB(
  ctx: CanvasRenderingContext2D,
  r: number,
  g: number,
  b: number,
) {
  r = linearToGamma(r);
  g = linearToGamma(g);
  b = linearToGamma(b);

  ctx.fillStyle = `rgb(${n(r)},${n(g)},${n(b)})`;
}

export function writeColor(
  ctx: CanvasRenderingContext2D,
  color: Vec3, // each component is [0,1]
  j: number,
  i: number,
) {
  fillRGB(ctx, color.x, color.y, color.z);
  ctx.fillRect(i, j, 1, 1);
}

File: raytracer/color.go
package raytracer

import (
	"image/color"
	"math"
	"raytracer/math3"
)

func colorToInt(val float64) uint8 {
	return uint8(math.Floor(val))
}

func linearToGamma(val float64) float64 {
	if val > 0 {
		return math.Sqrt(val)
	}
	return 0
}

func convertPixel(pixel math3.Vec3) color.Color {
	return color.RGBA{
		R: colorToInt(linearToGamma(pixel[0]) * 255),
		G: colorToInt(linearToGamma(pixel[1]) * 255),
		B: colorToInt(linearToGamma(pixel[2]) * 255),
		A: 255,
	}
}
